Выполнить ТЗ спринта №9

SergeyLeshchev opened the pull request:
1. Реализовать API, серверное приложение с запросами к менеджеру
2. Добавить 6 классов обработчиков(HttpHandler)
3. Добавить 3 адаптера для работы с JSON
4. Добавить 2 класса исключений и их обработку
5. Переместить метод main в класс HttpTaskServer
6. Изменить логику приложения, задачи с пересечением не добавляются
7. Добавить новые юнит-тесты для покрытия нового функционала

Commit:
Выполнить ТЗ спринта №9

1. Реализовать API, серверное приложение с запросами к менеджеру
2. Добавить 6 классов обработчиков(HttpHandler)
3. Добавить 3 адаптера для работы с JSON
4. Добавить 2 класса исключений и их обработку
5. Переместить метод main в класс HttpTaskServer
6. Изменить логику приложения, задачи с пересечением не добавляются
7. Добавить новые юнит-тесты для покрытия нового функционала

Commit:
Правки в код-стайл

Commit:
Убрать задержку 1 секунду ожидания сервера

Ivanov-s-tmn commented on .idea/java-kanban.iml:
Директорию .idea и файлы с расширением .iml стоит добавить в исключения gitignore. Они не нужны в репозитории и генерируются локально идеей

Ivanov-s-tmn commented on src/httphandler/BaseHttpHandler.java:
```diff
@@ -0,0 +1,79 @@
+
+public class BaseHttpHandler implements HttpHandler {
+    protected TaskManager taskManager;
+    Gson gson;
+    Task task;
+    Epic epic;
+    Subtask subtask;
```
Не забывай про инкапсуляцию и модификаторы доступа

Ivanov-s-tmn commented on src/httphandler/BaseHttpHandler.java:
```diff
@@ -0,0 +1,79 @@
+    @Override
+    public void handle(HttpExchange exchange) throws IOException {
+
+    }
```
Если сделать класс абстрактным, то можно не реализовывать этот метод, а переопределять его уже в наследниках

Ivanov-s-tmn commented on src/httphandler/BaseHttpHandler.java:
```diff
@@ -0,0 +1,79 @@
+            case "GET prioritized" -> Endpoint.GET_PRIORITIZED;
+            default -> Endpoint.UNKNOWN;
+        };
+    }
```
Этот метод стоит разбить и реализовать в хендлерах, чтобы он возвращал только нужные эндпоинты в зависимости от хендлера. Иначе это может сильно разрастись
Плюс, метод сделай приватным, методы ниже аналогично (protected, чтобы наследовались)

Ivanov-s-tmn commented on src/httphandler/EpicHttpHandler.java:
```diff
@@ -0,0 +1,42 @@
+
+import java.io.IOException;
+
+public class EpicHttpHandler extends BaseHttpHandler implements HttpHandler {
```
Имплементировать интерфейс `HttpHandler` не нужно

Ivanov-s-tmn commented on src/httphandler/HttpTaskServer.java:
```diff
@@ -0,0 +1,61 @@
+        System.out.println("Сервер остановлен");
+    }
+
+    public TaskManager getTaskManager() {
+        return taskManager;
+    }
+
+    public void setTaskManager(TaskManager taskManager) {
+        this.taskManager = taskManager;
+    }
```
А эти методы зачем?

Ivanov-s-tmn commented on src/httphandler/HistoryHttpHandler.java:
```diff
@@ -0,0 +1,19 @@
+
+    @Override
+    public void handle(HttpExchange exchange) throws IOException {
+        writeResponse(exchange, gson.toJson(taskManager.getHistory()), 200);
```
А если придет не тот метод? Этот момент нужно обработать. В хендлере prioritizedTasks тоже

Ivanov-s-tmn commented on src/httphandler/TaskAdapter.java:
Для тасок не нужны токены, по идее нужно реализовать только для полей с типом LDT и Duration

SergeyLeshchev replied:
Я сделал токен для тасок, потому что в момент, когда Gson сериализует и десериализует объекты и json, то все сабтаски и эпики становятся тасками и теряют часть своих полей, которые не свойственны Task. У меня не получилось реализовать другой вариант, чтобы коллекции передавали внутри себя таски, сабтаски и эпики и сохраняли все свойства...

Ivanov-s-tmn replied:
Понял, в таком случае окей)

Ivanov-s-tmn commented on src/httphandler/HttpTaskServer.java:
```diff
@@ -0,0 +1,61 @@
+
+    public Gson getGson() {
+        return gson;
+    }
```
Можно вынести в этот метод инициализацию и настройку gson вместо конструктора, сделать статиком и использовать где нужно

Commit:
Сделать работу над ошибками в ТЗ №9

1. Убрать лишние файлы из коммита
2. Проработать инкапсуляцию и модификаторы доступа к полям
3. Сделать класс BaseHttpHandler абстрактным
4. Распределить код метода getEndpoint() между обработчиками
5. Исправить методы handle() в обработчиках history и prioritized
6. Сделать gson статичным и изменить его инициализацию
7. Удалить лишний код и некоторые другие незначительные исправления
